src/MonoMod.RuntimeDetour/DetourManager.Shared.cs

```diff
@@ -190,10 +192,15 @@ internal sealed class DepGraph<TNode> {
 
         #region SyncInfo
         internal class DetourSyncInfo {
+            public MethodBase? SyncProxy;
             public int ActiveCalls;
             public int UpdatingThread;
 
-            public void WaitForChainUpdate() {
+            public bool WaitForChainUpdate() {
+                // If this is a nested call, allow the call to continue without waiting for the update to avoid deadlocks
+                if (Volatile.Read(ref ActiveCalls) > 1 && DetermineThreadCallDepth() > 1)
+                    return true;
+
                 _ = Interlocked.Decrement(ref ActiveCalls);
 
                 if (UpdatingThread == EnvironmentEx.CurrentManagedThreadId) {
```

Comment:
Isn't this now redundant?

Reply:
This code path is intended to catch the edge case where one thread calls the method, another thread tries to add/remove a hook, and then the first thread tries to perform a nested call - without this edge case this would end up deadlocking, as the nested call will block on the first thread, which will wait for the second thread's original call to return.